cgp/genome.py

```diff
@@ -134,6 +137,78 @@ def __repr__(self) -> str:
         s += ")"
         return s
 
+    def determine_permissible_value_per_gene(self, gene_idx: int, region_idx: int) -> np.array:
+
+        if self._is_input_region(region_idx):
+            return self._determine_permissible_values_input_gene(gene_idx, region_idx)
+
+        elif self._is_hidden_region(region_idx):
+            return self._determine_permissible_values_hidden_gene(gene_idx, region_idx)
+
+        elif self._is_output_region(region_idx):
+            return self._determine_permissible_values_output_gene(gene_idx)
+
+        else:
+            assert False  # should never be reached
+
+    def determine_permissible_values(self) -> List[np.array]:
+        """Determine permissible values for every gene.
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        ----------
+        permissible_values
+            List[numpy array]: List of numpy arrays over permissible values for every gene
+        """
+        permissible_values = []
+        for gene_idx in range(self._n_genes):
+            region_idx = gene_idx // self._length_per_region
+            permissible_values_per_gene = self.determine_permissible_value_per_gene(
+                gene_idx, region_idx
+            )
+            permissible_values.append(permissible_values_per_gene)
+        return permissible_values
+
+    def _determine_permissible_values_input_gene(self, gene_idx: int, region_idx: int) -> np.array:
+
+        if region_idx == 0:
+            region_idx += self._length_per_region  # avoid issues with x % 0
+
+        if gene_idx % (region_idx * self._length_per_region) == 0:
+            return np.array(self._id_input_node)
+        else:
+            return np.array(self._id_unused_gene)
+
+    def _determine_permissible_values_hidden_gene(
+        self, gene_idx: int, region_idx: int
+    ) -> np.array:
+
+        if self._is_function_gene(gene_idx):
+            permissible_values = np.arange(len(self._primitives._primitives))
+
+        elif self._is_hidden_input_gene(gene_idx, region_idx):
+            permissible_values = np.array(self._permissible_inputs(region_idx))
+
+        else:
+            assert False
+        return permissible_values
+
+    def _determine_permissible_values_output_gene(self, gene_idx: int,) -> np.array:
+        input_index = (
+            gene_idx % self._length_per_region
+        )  # assumes that the second gene in all output regions is the index of the input
+        if input_index == 0:
+            permissible_values = np.array(self._id_output_node)
+        elif input_index == 1:
+            permissible_values = np.array(self._permissible_inputs_for_output_region())
```

Comment:
```suggestion
            return np.array(self._permissible_inputs_for_output_region())
```
always return early when possible ;)